std/algorithm/comparison.d

```diff
@@ -1567,3 +1567,203 @@ unittest
+*/
+auto ref either(Ts...)(ref Ts a) if (a.length >= 1 && allSame!Ts)
+{
+    static if (Ts.length == 1)
+    {
+        return a[0];
+    }
+    else
+    {
+        return a[0] ? a[0] : either(a[1 .. $]); // recurse
+    }
+}
+
+@safe pure unittest
+{
+    immutable p = 1, q = 2;
+    auto pq = either(p, q);
+    assert(pq == 1);
+
+    assert(either(3) == 3);
+    assert(either(3, 4) == 3);
+    assert(either(0, 4) == 4);
+    assert(either(0, 0) == 0);
+    assert(either("", "a") == "");
+    string s = null;
+    assert(either(s, "a") == "a");
+    assert(either("a", "") == "a");
+    immutable a2 = [1, 2];
+    assert(either(a2) == a2);
+    assert(either([0, 1], [1, 2]) == [0, 1]);
+    assert(either([0, 1], [1]) == [0, 1]);
+    assert(either("a", "b") == "a");
+
+    int x = 1, y = 2;
+    either(x, y) = 3;
+    assert(x == 3);
+    assert(y == 2);
+}
+
+/** Returns: Last Argument if all arguments implicitly bool-convert to true
+    otherwise CommonType!T.init.
+
+    Similar to behaviour and operator in dynamic languages such as of Lisp's
+    (and a...) and Python's a and ....
+
+    TODO Is inout Conversion!T the correct return value?
+*/
+CommonType!T every(T...)(lazy T a) if (T.length >= 1)
+{
+    auto a0 = a[0]();           // evaluate only once
+    static if (T.length == 1)
+    {
+        return a0;
+    }
+    else
+    {
+        return a0 ? every(a[1 .. $]) : CommonType!T.init; // recurse
+    }
+}
+
+@safe pure unittest
+{
+    assert(every(3) == 3);
+    assert(every(3, 4) == 4);
+    assert(every(0, 4) == 0);
+    assert(every(0, 0) == 0);
+    assert(every([0, 1], [1, 2]) == [1, 2]);
+    assert(every([0, 1], [1]) == [1]);
+    assert(every("a", "b") == "b");
+    assert(every("", "b") == "b");
+    assert(every(cast(string)null, "b") == cast(string)null);
+}
+
+version(none) // WARNING disabled because I don't see any use of this for.
+{
+    /** This overload enables, when possible, lvalue return.
+     */
+    auto ref every(T...)(ref T a) if (T.length >= 1 && allSame!T)
+    {
+        static if (T.length == 1)
+        {
+            return a[0];
+        }
+        else
+        {
+            return a[0] ? every(a[1 .. $]) : a[0]; // recurse
+        }
+    }
+
+    unittest
+    {
+        immutable p = 1, q = 2;
+        assert(every(p, q) == 2);
+
+        int x = 1, y = 2;
+        every(x, y) = 3;
+        assert(x == 1);
+        assert(y == 3);
+    }
+}
+
+/** Evaluate all $(D parts) possibly digesting $(D whole).
+    If all values of $(D parts) implicitly convert to bool true return the
+    values as an array, otherwise restore whole and return null.
+*/
+CommonType!T[] tryEvery(S, T...)(ref S whole,
+                                 lazy T parts) if (T.length >= 1)
+{
+    auto wholeBackup = whole;
+    bool all = true;
+    alias R = typeof(return);
+    R results;
+    foreach (result; parts) // evaluate each part
+    {
+        if (result)
+        {
+            results ~= result;
```

Comment:
These array appends would be highly problematic for the performance of a parser...